README.md
# ExIban

Library for manipulating and validating IBAN account numbers.

## Installation

The package can be installed as:

  1. Add exiban to your list of dependencies in `mix.exs`:

        def deps do
          [{:exiban, "~> 0.0.1"}]
        end


  2. Run `mix deps.get` in your console to fetch from Hex


## Usage
    iex> ExIban.valid?("GB82 WEST 1234 5698 7654 32")
    true

    iex> ExIban.valid?("GB82 WEST 999 9999 9999")
    false

### Ecto validator

    defmodule User do
      use Ecto.Schema
      use Ecto.Model

      schema "users" do
        field :email, :string
        field :iban, :string
      end

      def changeset(user, params \\ :empty) do
        user
        |> cast(params, ~w(iban email), ~w())
        |> ExIban.EctoValidator.validate_iban(:iban)
      end
    end

## Documentation
Hosted on [http://hexdocs.pm/exiban/ExIban.html](http://hexdocs.pm/exiban/ExIban.html)

## Author
Krzysztof Kempiński

ExIban is released under the [MIT License](https://github.com/appcues/exsentry/blob/master/LICENSE.txt).


CHANGELOG.md
# Changelog

## v0.0.4

* ecto validator returns tuple in format `{field, error_msg}`

## v0.0.3

* add `format_compact` function

## v0.0.2

* Ecto validator

## v0.0.1

* basic functionality


test/exiban_test.exs
defmodule ExIbanTest do
  use ExUnit.Case
  doctest ExIban

end


test/parser_test.exs
defmodule ExIban.ParserTest do
  use ExUnit.Case
  doctest ExIban.Parser
end


test/validators_test.exs
defmodule ExIban.ValidatorsTest do
  use ExUnit.Case
  doctest ExIban.Validators
end


lib/rules.ex
defmodule ExIban.Rules do
  @moduledoc """
  Set of rules to check during validation.
  """

  @doc """
  Map of rules. Key represents country code.
  """

  @spec rules :: %{bitstring => %{bitstring => integer, bitstring => bitstring}}
  def rules do
    # https://github.com/iulianu/iban-tools/blob/master/lib/iban-tools/rules.yml
    %{"AD" => %{"length" => 24, "bban_pattern" => "\\d{8}[A-Z0-9]{12}"},
      "AE" => %{"length" => 23, "bban_pattern" => "\\d{19}"},
      "AL" => %{"length" => 28, "bban_pattern" => "\\d{8}[A-Z0-9]{16}"},
      "AT" => %{"length" => 20, "bban_pattern" => "\\d{16}"},
      "BA" => %{"length" => 20, "bban_pattern" => "\\d{16}"},
      "BE" => %{"length" => 16, "bban_pattern" => "\\d{12}"},
      "BG" => %{"length" => 22, "bban_pattern" => "[A-Z]{4}\\d{6}[A-Z0-9]{8}"},
      "BH" => %{"length" => 22, "bban_pattern" => "[A-Z]{4}[A-Z0-9]{14}"},
      "CH" => %{"length" => 21, "bban_pattern" => "\\d{5}[A-Z0-9]{12}"},
      "CY" => %{"length" => 28, "bban_pattern" => "\\d{8}[A-Z0-9]{16}"},
      "CZ" => %{"length" => 24, "bban_pattern" => "\\d{20}"},
      "DE" => %{"length" => 22, "bban_pattern" => "\\d{18}"},
      "DK" => %{"length" => 18, "bban_pattern" => "\\d{14}"},
      "DO" => %{"length" => 28, "bban_pattern" => "[A-Z]{4}\\d{20}"},
      "EE" => %{"length" => 20, "bban_pattern" => "\\d{16}"},
      "ES" => %{"length" => 24, "bban_pattern" => "\\d{20}"},
      "FI" => %{"length" => 18, "bban_pattern" => "\\d{14}"},
      "FO" => %{"length" => 18, "bban_pattern" => "\\d{14}"},
      "FR" => %{"length" => 27, "bban_pattern" => "\\d{10}[A-Z0-9]{11}\\d{2}"},
      "GB" => %{"length" => 22, "bban_pattern" => "[A-Z]{4}\\d{14}"},
      "GE" => %{"length" => 22, "bban_pattern" => "[A-Z]{2}\\d{16}"},
      "GI" => %{"length" => 23, "bban_pattern" => "[A-Z]{4}[A-Z0-9]{15}"},
      "GL" => %{"length" => 18, "bban_pattern" => "\\d{14}"},
      "GR" => %{"length" => 27, "bban_pattern" => "\\d{7}[A-Z0-9]{16}"},
      "HR" => %{"length" => 21, "bban_pattern" => "\\d{17}"},
      "HU" => %{"length" => 28, "bban_pattern" => "\\d{24}"},
      "IE" => %{"length" => 22, "bban_pattern" => "[A-Z]{4}\\d{14}"},
      "IL" => %{"length" => 23, "bban_pattern" => "\\d{19}"},
      "IS" => %{"length" => 26, "bban_pattern" => "\\d{22}"},
      "IT" => %{"length" => 27, "bban_pattern" => "[A-Z]\\d{10}[A-Z0-9]{12}"},
      "KW" => %{"length" => 30, "bban_pattern" => "[A-Z]{4}\\d{22}"},
      "KZ" => %{"length" => 20, "bban_pattern" => "[0-9]{3}[A-Z0-9]{13}"},
      "LB" => %{"length" => 28, "bban_pattern" => "\\d{4}[A-Z0-9]{20}"},
      "LI" => %{"length" => 21, "bban_pattern" => "\\d{5}[A-Z0-9]{12}"},
      "LT" => %{"length" => 20, "bban_pattern" => "\\d{16}"},
      "LU" => %{"length" => 20, "bban_pattern" => "\\d{3}[A-Z0-9]{13}"},
      "LV" => %{"length" => 21, "bban_pattern" => "[A-Z]{4}[A-Z0-9]{13}"},
      "MC" => %{"length" => 27, "bban_pattern" => "\\d{10}[A-Z0-9]{11}\\d{2}"},
      "ME" => %{"length" => 22, "bban_pattern" => "\\d{18}"},
      "MK" => %{"length" => 19, "bban_pattern" => "\\d{3}[A-Z0-9]{10}\\d{2}"},
      "MR" => %{"length" => 27, "bban_pattern" => "\\d{23}"},
      "MT" => %{"length" => 31, "bban_pattern" => "[A-Z]{4}\\d{5}[A-Z0-9]{18}"},
      "MU" => %{"length" => 30, "bban_pattern" => "[A-Z]{4}\\d{19}[A-Z]{3}"},
      "NL" => %{"length" => 18, "bban_pattern" => "[A-Z]{4}\\d{10}"},
      "NO" => %{"length" => 15, "bban_pattern" => "\\d{11}"},
      "PL" => %{"length" => 28, "bban_pattern" => "\\d{8}[A-Z0-9]{16}"},
      "PT" => %{"length" => 25, "bban_pattern" => "\\d{21}"},
      "RO" => %{"length" => 24, "bban_pattern" => "[A-Z]{4}[A-Z0-9]{16}"},
      "RS" => %{"length" => 22, "bban_pattern" => "\\d{18}"},
      "SA" => %{"length" => 24, "bban_pattern" => "\\d{2}[A-Z0-9]{18}"},
      "SE" => %{"length" => 24, "bban_pattern" => "\\d{20}"},
      "SI" => %{"length" => 19, "bban_pattern" => "\\d{15}"},
      "SK" => %{"length" => 24, "bban_pattern" => "\\d{20}"},
      "SM" => %{"length" => 27, "bban_pattern" => "[A-Z]\\d{10}[A-Z0-9]{12}"},
      "TN" => %{"length" => 24, "bban_pattern" => "\\d{20}"},
      "TR" => %{"length" => 26, "bban_pattern" => "\\d{5}[A-Z0-9]{17}"},
      "UA" => %{"length" => 29, "bban_pattern" => "\\d{25}"}}
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :exiban, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:exiban, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/validators.ex
defmodule ExIban.Validators do
  @moduledoc """
  Set of validation rules to perform on checking IBAN account number.
  """

  import ExIban.Rules
  import ExIban.Parser

  @doc """
  Performs validation checks:
  - length greater or equal to 5 (:too_short)
  - only ilegal chars used [A-Z0-9] (bad_chars)
  - known country code (:unknown_country_code)
  - correct length for given country (:bad_length)
  - correct bban format for given country (:bad_format)
  - checking digits (:bad_check_digits)

  ## Examples

      iex> ExIban.Validators.issues("GB82")
      {[:too_short], nil}

      iex> ExIban.Validators.issues("GB_+)*")
      {[:bad_chars, :bad_length, :bad_format, :bad_check_digits], {"GB", "_+", ")*", 6, "GB_+)*"}}

      iex> ExIban.Validators.issues("GB82 WEST 1234 5698 7654 32")
      {[], {"GB", "82", "WEST12345698765432", 22, "GB82WEST12345698765432"}}
  """

  @spec issues(binary) :: {list, {bitstring, bitstring, bitstring, integer, bitstring} | nil}
  def issues(iban) when byte_size(iban) < 5, do: {[:too_short], nil}
  def issues(iban) do
    iban
    |> parse
    |> do_validation
  end

  defp do_validation(iban) do
    iban
    |> check_chars
    |> check_country_code
    |> check_length
    |> check_format
    |> check_digits
  end

  defp check_chars({_, _, _, _, iban} = parsed_iban) do
    cond do
      not Regex.match?(~r/^[A-Z0-9]+$/, iban) -> {[:bad_chars], parsed_iban}
      true                                    -> {[], parsed_iban}
    end
  end

  defp check_country_code({errors, {country_code, _, _, _, _} = parsed_iban}) do
    cond do
      rules |> Map.get(country_code) |> is_nil ->
        {errors ++ [:unknown_country_code], parsed_iban}
      true -> {errors, parsed_iban}
    end
  end

  defp check_length({errors,
    {country_code, _, _, iban_length, _} = parsed_iban}) do
    cond do
      rules |> Map.get(country_code, %{}) |> Map.get("length") != iban_length ->
        {errors ++ [:bad_length], parsed_iban}
      true -> {errors, parsed_iban}
    end
  end

  defp check_format({errors,
    {country_code, _, bban, _, _} = parsed_iban}) do
    {:ok, reg} = rules
                  |> Map.get(country_code, %{})
                  |> Map.get("bban_pattern", "")
                  |> Regex.compile
    cond do
      not Regex.match?(reg, bban) -> {errors ++ [:bad_format], parsed_iban}
      true                        -> {errors, parsed_iban}
    end
  end

  defp check_digits({errors,
    {country_code, check_digits, bban, _, _} = parsed_iban}) do
    chars = String.to_char_list(bban <> country_code <> check_digits)
    numbers = for byte <- chars, into: [] do
      case byte do
        byte when byte in 48..57 -> List.to_string([byte])
        byte when byte in 65..90 -> Integer.to_string(byte - 55)
        _ -> ""
      end
    end

    cond do
      numbers |> Enum.join |> String.to_integer |> rem(97) != 1 ->
        {errors ++ [:bad_check_digits], parsed_iban}
      true -> {errors, parsed_iban}
    end
  end
end


lib/exiban/parser.ex
defmodule ExIban.Parser do
  @moduledoc """
  Parse IBAN bank account as a string and prepare nice tuple.
  """

  @doc """
  Parse IBAN and return tuple
  {country_code, check_digits, bban, length of iban, normalized iban}

  ## Examples

      iex> ExIban.Parser.parse("  GB82 WEST 1234 5698 765432  ")
      {"GB", "82", "WEST12345698765432", 22, "GB82WEST12345698765432"}
  """

  @spec parse(binary) :: {bitstring, bitstring, bitstring, integer, bitstring}
  def parse(iban) do
    iban = iban |> normalize

    {
      String.slice(iban, 0, 2),
      String.slice(iban, 2, 2),
      String.slice(iban, 4, String.length(iban) - 4),
      String.length(iban),
      iban
    }
  end

  defp normalize(iban) do
    iban
    |> String.replace(~r/\s+/, "")
    |> String.strip
    |> String.upcase
  end
end


lib/exiban/ecto_validator.ex
defmodule ExIban.EctoValidator do
  @moduledoc """
  Validator for Ecto
  """

  @doc """
  Validate Ecto changeset. Returns changeset or tuple in format `{field, error_msg}` which can be useful for Phoenix.

  ## Examples

      defmodule User do
        use Ecto.Schema
        use Ecto.Model

        schema "users" do
          field :email, :string
          field :iban, :string
        end

        def changeset(user, params \\\\ :empty) do
          user
          |> cast(params, ~w(iban email), ~w())
          |> ExIban.EctoValidator.validate_iban(:iban)
        end
      end
  """

  @spec validate_iban(map, bitstring) :: map | {:error, tuple}
  def validate_iban(changeset, field) do
    %{changes: changes, errors: errors} = changeset

    iban = Map.get(changes, field)
    case ExIban.validate(iban) do
      {:error, new_errors} ->
        new_errors = new_errors
                      |> Enum.map(&Atom.to_string/1)
                      |> Enum.map(&(String.replace(&1, "_", " ")))
                      |> Enum.map(&({field, &1}))
        %{changeset | errors: new_errors ++ errors, valid?: false}
      :ok -> changeset
    end
  end
end


lib/exiban.ex
defmodule ExIban do
  @moduledoc """
  Validate and manipulate IBAN account number.
  """

  import ExIban.Parser
  import ExIban.Validators

  @doc """
  Perform validation rules on IBAN.

  ## Examples

      iex> ExIban.validate("GB82 WEST 1234 5698 7654 32")
      :ok

      iex> ExIban.validate("GB82 WEST 999 9999 9999")
      { :error, [:bad_length, :bad_format, :bad_check_digits] }
  """

  @spec validate(bitstring) :: :ok | {:error, list}
  def validate(iban) do
    iban
    |> issues
    |> prepare_output
  end

  @doc """
  Checks if IBAN is valid.

  ## Examples

      iex> ExIban.valid?("GB82 WEST 1234 5698 7654 32")
      true

      iex> ExIban.valid?("GB82 WEST 999 9999 9999")
      false
  """

  @spec valid?(bitstring) :: true | false
  def valid?(iban) do
    iban
    |> issues
    |> do_valid?
  end

  @doc """
  Returns country code extracted from IBAN.

  ## Examples

      iex> ExIban.country_code("GB82 WEST 1234 5698 7654 32")
      "GB"
  """

  @spec country_code(bitstring) :: bitstring
  def country_code(iban) do
    {country_code, _, _, _, _} = parse(iban)
    country_code
  end

  @doc """
  Returns bban extracted from IBAN.

  ## Examples

      iex> ExIban.bban("GB82 WEST 1234 5698 7654 32")
      "82"
  """

  @spec bban(bitstring) :: bitstring
  def bban(iban) do
    {_, bban, _, _, _} = parse(iban)
    bban
  end

  @doc """
  Format and pretify IBAN to be more readable.

  ## Examples

      iex> ExIban.format(" GB 82     WEST 1234 5698765432  ")
      "GB82 WEST 1234 5698 7654 32"
  """

  @spec format(bitstring) :: bitstring
  def format(iban) do
    {_, _, _, _, formatted} = parse(iban)
    formatted
    |> String.replace(~r/(.{4})/, "\\1 ")
    |> String.strip
  end

  @doc """
  Format and IBAN to compacted version.

  ## Examples

      iex> ExIban.format_compact(" GB 82     WEST 1234 5698765432  ")
      "GB82WEST12345698765432"
  """

  @spec format_compact(bitstring) :: bitstring
  def format_compact(iban) do
    {_, _, _, _, formatted} = parse(iban)
    formatted
    |> String.strip
  end

  defp prepare_output({[], _}), do: :ok
  defp prepare_output({issues, _}) do
    {:error, issues}
  end

  defp do_valid?({[], _}), do: true
  defp do_valid?(_), do: false
end
